receiver/splunkhecreceiver/receiver.go

```diff
@@ -0,0 +1,319 @@
+			}
+		}()
+	})
+
+	return err
+}
+
+// StopMetricsReception tells the receiver that should stop reception,
+// giving it a chance to perform any necessary clean-up.
+func (r *splunkReceiver) Shutdown(context.Context) error {
+	r.Lock()
+	defer r.Unlock()
+
+	err := componenterror.ErrAlreadyStopped
+	r.stopOnce.Do(func() {
+		err = r.server.Close()
+	})
+	return err
+}
+
+func (r *splunkReceiver) handleReq(resp http.ResponseWriter, req *http.Request) {
+	transport := "http"
+	if r.config.TLSSetting != nil {
+		transport = "https"
+	}
+	ctx := obsreport.ReceiverContext(req.Context(), r.config.Name(), transport, r.config.Name())
+	ctx = obsreport.StartMetricsReceiveOp(ctx, r.config.Name(), transport)
+
+	if req.Method != http.MethodPost {
+		r.failRequest(ctx, resp, http.StatusBadRequest, invalidMethodRespBody, nil)
+		return
+	}
+
+	if req.Header.Get(httpContentTypeHeader) != jsonContentType {
+		r.failRequest(ctx, resp, http.StatusUnsupportedMediaType, invalidContentRespBody, nil)
+		return
+	}
+
+	encoding := req.Header.Get(httpContentEncodingHeader)
+	if encoding != "" && encoding != gzipEncoding {
+		r.failRequest(ctx, resp, http.StatusUnsupportedMediaType, invalidEncodingRespBody, nil)
+		return
+	}
+
+	bodyReader := req.Body
+	if encoding == gzipEncoding {
+		var err error
+		bodyReader, err = gzip.NewReader(bodyReader)
+		if err != nil {
+			r.failRequest(ctx, resp, http.StatusBadRequest, errGzipReaderRespBody, err)
+			return
+		}
+	}
+
+	body, err := ioutil.ReadAll(bodyReader)
+	if err != nil {
+		r.failRequest(ctx, resp, http.StatusBadRequest, errReadBodyRespBody, err)
+		return
+	}
+	if len(body) == 0 {
+		resp.Write(okRespBody)
+		return
+	}
+
+	messagesReceived := 0
+	dec := json.NewDecoder(bytes.NewReader(body))
+
+	var decodeErr error
+	// while the array contains values
+	for dec.More() {
+		var msg splunk.Metric
+		// decode an array value (Message)
+		err = dec.Decode(&msg)
+		if err != nil {
+			r.failRequest(ctx, resp, http.StatusBadRequest, errUnmarshalBodyRespBody, err)
+			return
+		}
+		messagesReceived += len(msg.GetValues())
+
+		md, _ := SplunkHecToMetricsData(r.logger, &msg)
+
+		if r.config.AccessTokenPassthrough {
+			if accessToken := req.Header.Get("Splunk"); accessToken != "" {
+				if md.Resource == nil {
+					md.Resource = &resourcepb.Resource{}
+				}
+				if md.Resource.Labels == nil {
+					md.Resource.Labels = make(map[string]string, 1)
+				}
+				md.Resource.Labels["com.splunk.hec.access_token"] = accessToken
+			}
+		}
+		m := pdatautil.MetricsFromMetricsData([]consumerdata.MetricsData{*md})
+		decodeErr = r.nextConsumer.ConsumeMetrics(ctx, m)
+		if decodeErr != nil {
+			break
+		}
+	}
+
+	obsreport.EndMetricsReceiveOp(
+		ctx,
+		typeStr,
+		messagesReceived,
+		messagesReceived,
+		decodeErr)
+	if decodeErr != nil {
+		r.failRequest(ctx, resp, http.StatusInternalServerError, errNextConsumerRespBody, err)
+	} else {
+		resp.WriteHeader(http.StatusAccepted)
+		resp.Write(okRespBody)
+	}
```

Comment:
Why not do this at point of decode failure instead of after EndMetricsReceiveOp?

Reply:
Looks like `decodeErr` is used in `EndMetricsReceiveOp`, to pass status back.